lib/IDE/ArgumentCompletion.cpp

```diff
@@ -175,6 +175,24 @@ void ArgumentTypeCheckCompletionCallback::sawSolutionImpl(const Solution &S) {
   if (!isExpressionResultTypeUnconstrained(S, ParentCall)) {
     ExpectedCallType = getTypeForCompletion(S, ParentCall);
   }
+  // If we recorded a fix to allow a type mismatch for the ParentCall,
+  // use the type `ParentCall` should have had as the expected call type.
+  // For example, in the following the call to `bar` gets assigned a type
+  // of `String` but we want to report `Int` as the expected call type.
+  //
+  // func foo(_ x: Int) {}
+  // func bar() -> String {}
+  // foo(bar(#^COMPLETE^#))
+  for (auto Fix : S.Fixes) {
+    if (auto CM = dyn_cast<AllowArgumentMismatch>(Fix)) {
+      if (simplifyLocatorToAnchor(CM->getLocator()) == ASTNode(ParentCall)) {
+        ExpectedCallType = CM->getToType();
```

Comment:
Why can't we merge two if conditions into one using `&&` operator. This might improve readability

Reply:
I don’t think so because one of them is binding a variable (`auto CM`) and the other one is a boolean check.

Reply:
Technically we could now do (as of C++17) `if (auto CM = dyn_cast<AllowArgumentMismatch>(Fix); simplifyLocatorToAnchor(CM->getLocator()) == ASTNode(ParentCall))`. But we don't use this pattern anywhere else in the compiler and IMO we probably want to agree in general that it's a pattern we want to use first.

Reply:
The part of the condition after `;` still gets evaluated even if `CM` is `nullptr`. Which means that it would need to be `if (auto CM = dyn_cast<AllowArgumentMismatch>(Fix); CM && simplifyLocatorToAnchor(CM->getLocator()) == ASTNode(ParentCall))` which I find sufficiently ugly to not use.